src/ir/possible-contents.cpp

```diff
@@ -0,0 +1,1652 @@
+    auto* parent = std::get<ExpressionLocation>(getLocation(parentIndex)).expr;
+
+#if defined(POSSIBLE_CONTENTS_DEBUG) && POSSIBLE_CONTENTS_DEBUG >= 2
+    std::cout << "  special, parent:\n" << *parent << '\n';
+#endif
+
+    if (auto* get = parent->dynCast<StructGet>()) {
+      // |child| is the reference child of a struct.get.
+      assert(get->ref == child);
+      readFromData(get->ref->type.getHeapType(), get->index, contents, get);
+    } else if (auto* set = parent->dynCast<StructSet>()) {
+      // |child| is either the reference or the value child of a struct.set.
+      assert(set->ref == child || set->value == child);
+      writeToData(set->ref, set->value, set->index);
+    } else if (auto* get = parent->dynCast<ArrayGet>()) {
+      assert(get->ref == child);
+      readFromData(get->ref->type.getHeapType(), 0, contents, get);
+    } else if (auto* set = parent->dynCast<ArraySet>()) {
+      assert(set->ref == child || set->value == child);
+      writeToData(set->ref, set->value, 0);
+    } else if (auto* cast = parent->dynCast<RefCast>()) {
+      assert(cast->ref == child);
+      flowRefCast(contents, cast);
+    } else {
+      // TODO: ref.test and all other casts can be optimized (see the cast
+      //       helper code used in OptimizeInstructions and RemoveUnusedBrs)
+      WASM_UNREACHABLE("bad childParents content");
+    }
+  }
+}
+
+void Flower::flowToTargetsAfterUpdate(LocationIndex locationIndex,
+                                      const PossibleContents& contents) {
+  // Send the new contents to all the targets of this location. As we do so,
+  // prune any targets that we do not need to bother sending content to in the
+  // future, to save space and work later.
+  auto& targets = getTargets(locationIndex);
+  targets.erase(std::remove_if(targets.begin(),
+                               targets.end(),
+                               [&](LocationIndex targetIndex) {
+#if defined(POSSIBLE_CONTENTS_DEBUG) && POSSIBLE_CONTENTS_DEBUG >= 2
+                                 std::cout << "  send to target\n";
+                                 dump(getLocation(targetIndex));
+#endif
+                                 return !updateContents(targetIndex, contents);
+                               }),
+                targets.end());
+
+  if (contents.isMany()) {
+    // We contain Many, and just called updateContents on our targets to send
+    // that value to them. We'll never need to send anything from here ever
+    // again, since we sent the worst case possible already, so we can just
+    // clear our targets vector. But we should have already removed all the
+    // targets in the above remove_if operation, since they should have all
+    // notified us that we do not need to send them any more updates.
+    assert(targets.empty());
+  }
+}
+
+void Flower::connectDuringFlow(Location from, Location to) {
+  auto newLink = LocationLink{from, to};
+  auto newIndexLink = getIndexes(newLink);
```

Comment:
Why do links need indices, again? I would expect each location to just store a vector of its successor location indices with no need to separately materialize links.

Reply:
Correct, each location just stores a vector of successor indexes. That's what the code
```cpp
// Add it to the |targets| vector.
```
does a few lines down. However, we also keep `links` around simply to be able to tell what links exist in the program. The comment on `links` explains that,
```cpp
  // All existing links in the graph. We keep this to know when a link we want
  // to add is new or not.
  std::unordered_set<IndexLink> links;
```
(Instead, we could scan through the `targets` vector to avoid adding a link that already exists, but that would take linear time.)

Reply:
I've never used a bloom filter before - interesting!